trl/trainer/simpo_trainer.py

```diff
@@ -0,0 +1,441 @@
+# SimPO Authors: Yu Meng, Mengzhou Xia, Danqi Chen
+# Official code: https://github.com/princeton-nlp/SimPO
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import random
+import warnings
+from collections import defaultdict
+from contextlib import nullcontext
+from functools import wraps
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from accelerate import PartialState
+from datasets import Dataset
+from torch.utils.data import DataLoader
+from transformers import AutoModelForCausalLM, DataCollator, PreTrainedModel, PreTrainedTokenizerBase, Trainer
+from transformers.trainer_callback import TrainerCallback
+from transformers.trainer_utils import EvalLoopOutput
+from transformers.utils import is_torch_fx_proxy
+
+from ..import_utils import is_peft_available, is_wandb_available
+from .simpo_config import SimPOConfig
+from .utils import (
+    DPODataCollatorWithPadding,
+    disable_dropout_in_model,
+    pad_to_length,
+    peft_module_casting_to_bf16,
+    trl_sanitze_kwargs_for_tagging,
+)
+
+from trl.trainer import CPOTrainer
+
+if is_peft_available():
+    from peft import PeftModel, get_peft_model, prepare_model_for_kbit_training
+
+
+if is_wandb_available():
+    import wandb
+
+
+class SimPOTrainer(CPOTrainer):
```

Comment:
I'd prefer `SimPOTrainer` to inherit `Trainer`. It's more in line with our single-file philosophy.

Reply:
Ah, I see I'm contradicting https://github.com/huggingface/trl/pull/1725#issuecomment-2189147362 in a way. 

What about this:

```python
    def __init__(
        self,
        model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
        args: Optional[SimPOConfig] = None,
        data_collator: Optional[DataCollator] = None,
        train_dataset: Optional[Dataset] = None,
        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
        tokenizer: Optional[PreTrainedTokenizerBase] = None,
        model_init: Optional[Callable[[], PreTrainedModel]] = None,
        callbacks: Optional[List[TrainerCallback]] = None,
        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
        peft_config: Optional[Dict] = None,
        compute_metrics: Optional[Callable[[EvalLoopOutput], Dict]] = None,
    ):
    super().__init__(
        mode=model,
        args=args
        data_collator=data_collator
        train_dataset=train_dataset
        eval_dataset=eval_dataset
        tokenizer=tokenizer
        model_init=model_init
        callbacks=callbacks
        optimizers=optimizers
        preprocess_logits_for_metrics=preprocess_logits_for_metrics
        peft_config=peft_config
        compute_metrics=compute_metrics
    )
    self.gamma_beta_ratio = args.gamma_beta_ratio
    self.sft_weight = args.sft_weight
```

Instead of the full init?

Reply:
Your comment makes sense. At this point, I see two solutions that both seem good to me:
- Rewrite a complete trainer
- Make a very light subclass

I feel we're in the middle of these two solutions here. So I think that even if the error messages are misleading (CPO trainer instead of SimPO), this won't be a problem for the user.